nupic/research/frameworks/vernon/mixins/centroid_context.py

```diff
@@ -0,0 +1,128 @@
+    """
+
+    def setup_experiment(self, config):
+        super().setup_experiment(config)
+
+        # Store batch size
+        self.batch_size = config.get("batch_size", 1)
+
+        # Tensor for accumulating each task's centroid vector
+        self.contexts = torch.zeros((0, self.model.input_size))
+        self.contexts = self.contexts.to(self.device)
+
+        # The following will point to the the 'active' context vector used to train on
+        # the current task
+        self.context_vector = None
+
+    def run_task(self):
+        self.train_loader.sampler.set_active_tasks(self.current_task)
+
+        # Construct a context vector by computing the centroid of all training examples
+        self.context_vector = compute_centroid(self.train_loader).to(self.device)
+        self.contexts = torch.cat((self.contexts, self.context_vector.unsqueeze(0)))
+
+        return super().run_task()
+
+    def train_epoch(self):
+        # TODO: take out constants in the call below. How do we determine num_labels?
+        train_dendrite_model(
+            model=self.model,
+            loader=self.train_loader,
+            optimizer=self.optimizer,
+            device=self.device,
+            criterion=self.error_loss,
+            share_labels=True,
+            num_labels=10,
+            context_vector=self.context_vector,
+            post_batch_callback=self.post_batch_wrapper,
+        )
+
+    def validate(self, loader=None):
+        if loader is None:
+            loader = self.val_loader
+
+        # TODO: take out constants in the call below
+        return evaluate_dendrite_model(model=self.model,
+                                       loader=loader,
+                                       device=self.device,
+                                       criterion=self.error_loss,
+                                       share_labels=True, num_labels=10,
+                                       infer_context_fn=infer_centroid(self.contexts))
+
+
+def compute_centroid(loader):
+    """
+    Returns the centroid vector of all samples iterated over in `loader`.
+    """
+    centroid_vector = torch.zeros([])
+    n_centroid = 0
+    for x, _ in loader:
+        if isinstance(x, list):
+            x = x[0]
+        x = x.flatten(start_dim=1)
+        n_x = x.size(0)
+
+        centroid_vector = centroid_vector.to(x.device)
+
+        centroid_vector = n_centroid * centroid_vector + n_x * x.mean(dim=0)
+        centroid_vector = centroid_vector / (n_centroid + n_x)
+        n_centroid += n_x
+
+    return centroid_vector
+
+
+def infer_centroid(contexts):
+    """
+    Returns a function that takes a batch of test examples and returns a 2D array where
+    row i gives the the centroid vector closest to the ith test example.
+    """
+
+    def _infer_centroid(data):
+        context = torch.cdist(contexts, data)
+        context = context.argmin(dim=0)
+        context = context.cpu()
```

Comment:
Why `.cpu()` here?

Reply:
I think this is leftover from some debugging I did weeks ago. I removed it, since it seems like everything runs without error if we omit the cpu call.